File: bloom-filter-simple.pl
#!/usr/bin/env perl
use strict;
use warnings;

package BloomFilter;

use Digest::MD5 qw(md5_hex);

sub new {
    my ($class, %args) = @_;
    my $size = $args{size} || 20;
    
    my $self = {
        size       => $size,
        num_hashes => $args{num_hashes} || 3,
        bit_array  => [ (0) x $size ],
    };
    return bless $self, $class;
}

sub add {
    my ($self, $item) = @_;
    for my $i (0 .. $self->{num_hashes} - 1) {
		 # Gemini used a simple hash function here, which is not necessary as Digeest::MD5 is part of the base Perl distribution. So we can use it directly.
		 my $index = md5_hex("$item-$i");
		 $index = hex(substr($index, 0, 8)) % $self->{size};
		 $self->{bit_array}->[$index] = 1;
    }
}

sub contains {
    my ($self, $item) = @_;
    for my $i (0 .. $self->{num_hashes} - 1) {
		 my $index = md5_hex("$item-$i");
		 $index = hex(substr($index, 0, 8)) % $self->{size};
        return 0 if !$self->{bit_array}->[$index];
    }
    return 1;
}

# --- Demonstration ---
package main;

my $bf = BloomFilter->new(size => 20, num_hashes => 3);

# Add items
$bf->add("apple");
$bf->add("banana");

# Check membership
print "apple: "  . ($bf->contains("apple")  ? "Found" : "Not Found") . "\n"; # Found
print "cherry: " . ($bf->contains("cherry") ? "Found" : "Not Found") . "\n"; # Not Found

File: bloom-filter-simple.py
import hashlib

class SimpleBloomFilter:
    def __init__(self, size=100, hash_count=3):
        self.size = size
        self.hash_count = hash_count
        self.bit_array = [0] * size

    def _indexes(self, item):
        # Generate multiple positions by salting the item string
        indexes = []
        for i in range(self.hash_count):
            salted = f"{item}-{i}".encode('utf-8')
            digest = hashlib.md5(salted).hexdigest()
            index = int(digest, 16) % self.size
            indexes.append(index)
        return indexes

    def add(self, item):
        for index in self._indexes(item):
            self.bit_array[index] = 1

    def __contains__(self, item):
        # Returns True if maybe in set, False if definitely not in set
        for index in self._indexes(item):
            if self.bit_array[index] == 0:
                return False
        return True

# Demonstration
bf = SimpleBloomFilter(size=50, hash_count=3)
bf.add("apple")
bf.add("banana")

print("apple:", "apple" in bf)    # True (possibly in set)
print("cherry:",  "cherry" in bf)   # False (definitely not in set)

